labrad/crypto.py

```diff
@@ -0,0 +1,134 @@
+    raise an exception.
+
+    Args:
+        hostname (string): The hostname we are trying to connect to with TLS.
+        cert_string (string): A certificate in PEM format. If provided, we
+            return CertificateOptions that will trust only the given cert
+            (assuming its canonical name attribute matches the given hostname).
+            If not provided, we will try to load a suitable certificate for
+            this hostname from our store of trusted self-signed certificates,
+            and otherwise fall back to just using the trust roots configured
+            for the OS to validate the server certificate.
+
+    Returns (twisted.internet.ssl.CertificateOptions):
+        Certificate suitable for use by TLS clients to connect to the given
+
+    Raises:
+        Exception: could not import pyOpenSSL package needed for TLS.
+    """
+    if not TLS:
+        raise Exception('Unable to connect to labrad with TLS. Please be sure '
+                        'to install pyOpenSSL.')
+    if cert_string is None:
+        cert_string = load_cert(hostname)
+    if cert_string is None:
+        logging.info('No certificate found for host "{}" in {}. Will use '
+                     'system certs to verify tls'.format(hostname, CERTS_PATH))
+        return ssl.optionsForClientTLS(hostname.decode('utf-8'))
+    else:
+        cert = ssl.Certificate.loadPEM(cert_string)
+        return ssl.optionsForClientTLS(hostname.decode('utf-8'), cert)
+
+
+def save_cert(hostname, cert_string):
+    """Save the given self-signed certificate as a trusted certificate.
+
+    Args:
+        hostname (string): The hostname for this certificate. Must match the
+            canonical name attribute (CN) in the certificate itself. We do not
+            check this here, but if it is not the case, TLS negotiation using
+            this cert will fail.
+        cert_string (string): Certificate in PEM format.
+    """
+    # sanity check that this is a valid PEM string
+    cert = ssl.Certificate.loadPEM(cert_string)
+
+    fname = hostname + '.cert'
+    full_name = os.path.join(CERTS_PATH, fname)
+    if not os.path.exists(CERTS_PATH):
+        os.makedirs(CERTS_PATH)
+    with open(full_name, 'w') as f:
+        f.write(cert_string)
+
+
+def load_cert(hostname):
+    """Load a trusted certificate for the given hostname.
+
+    Args:
+        hostname(string): The hostname for which we want a certificate.
+
+    Returns (string | None):
+        A certificate for the specified hostname as a PEM-encoded string. If no
+        certificate is found for the given hostname, we return None.
+    """
+    if not os.path.exists(CERTS_PATH):
+        return None
+    for fname in os.listdir(CERTS_PATH):
+        full_name = os.path.join(CERTS_PATH, fname)
+        if not os.path.isfile(full_name):
+            continue
+        host, ext = os.path.splitext(fname)
+        if host == hostname and ext == '.cert':
+            with open(full_name) as f:
```

Comment:
So certificates have to be in a file hostname.cert, one per file?  I know we create them that way, but if we want to allow people to drop certificate files in here we should support multiple certificates per file, and use the CN field of the certificate rather than the filename.